src/lib/IDBDAO.js

```diff
@@ -0,0 +1,377 @@
+    },
+
+    function serialize(obj) {
+      return foam.json.stringify(obj); //TODO: serialization
+    },
+
+    function openDB(cc) {
+      var indexedDB = window.indexedDB ||
+        window.webkitIndexedDB         ||
+        window.mozIndexedDB;
+
+      var request = indexedDB.open("FOAM:" + this.name, 1);
+      var self = this;
+
+      request.onupgradeneeded = (function(e) {
+        var store = e.target.result.createObjectStore(self.name);
+        for ( var i = 0; i < self.indicies.length; i++ ) {
+          store.createIndex(self.indicies[i][0], self.indicies[i][0], { unique: self.indicies[i][1] });
+        }
+      })
+
+      request.onsuccess = (function(e) {
+        cc(e.target.result);
+      })
+
+      request.onerror = function (e) {
+        console.log('************** failure', e);
+      };
+    },
+
+    function withStore(mode, fn) {
+      if ( mode !== 'readwrite' ) return this.withStore_(mode, fn);
+
+      var self = this;
+
+      if ( ! this.q_ ) {
+        var q = [fn];
+        this.q_ = q;
+        this.async(function() {
+          self.withStore_(mode, function(store) {
+            // console.log('q length: ', q.length);
+            if ( self.q_ == q ) self.q_ = undefined;
+            for ( var i = 0 ; i < q.length ; i++ ) q[i](store);
+          });
+        }, this.X)();
+      } else {
+        this.q_.push(fn);
+        // Diminishing returns after 10000 per batch
+        if ( this.q_.length == 10000 ) this.q_ = undefined;
+      }
+    },
+
+    function withStore_(mode, fn) {
+      var self = this;
+      if ( self.__TXN__[0] ) {
+        try {
+          fn.call(self, self.__TXN__[0]);
+          return;
+        } catch (x) {
+          self.__TXN__[0] = undefined;
+        }
+      }
+      self.openDB(function (db) { // TODO: memoize like in foam1
+        var tx = db.transaction([self.name], mode);
+        var os = tx.objectStore(self.name);
+        self.__TXN__[0] = os;
+        fn.call(self, os);
+      });
+    },
+
+    function put(value, sink) {
+      var self = this;
+      return new Promise(function(resolve, reject) {
+        self.withStore("readwrite", function(store) {
+          var request = store.put(self.serialize(value), value.id);
+          request.transaction.addEventListener(
+            'complete',
+            function(e) {
+              self.pub('on','put', value);
+              sink && sink.put && sink.put(value);
+              resolve(sink || value);
+            });
+          request.transaction.addEventListener(
+            'error',
+            function(e) {
+              sink && sink.error && sink.error('put', value, e); // TODO: err message
+              reject(self.IDBException.create({ id: value.id, error: e }));
+            });
+        });
+      });
+    },
+
+    function find(key) {
+      var self = this;
+
+      if ( foam.mlang.predicate.Expr.isInstance(key) ) {
```

Comment:
I don't think we should support finding with an expression any more.  It was rarely used, and not always clear.  If you really want to find based off an expression you can do limit(1).where().select() yourself.